ledger/participant-integration-api/src/main/scala/platform/store/cache/StateCache.scala

```diff
@@ -30,40 +33,39 @@ private[platform] case class StateCache[K, V](cache: Cache[K, V], registerUpdate
   def get(key: K)(implicit loggingContext: LoggingContext): Option[V] =
     cache.getIfPresent(key) match {
       case Some(value) =>
-        logger.debug(s"Cache hit for $key -> ${value.toString.take(100)}")
+        logger.debug(s"Cache hit for $key -> ${value.toString.take(1000)}")
         Some(value)
       case None =>
         logger.debug(s"Cache miss for $key ")
         None
     }
 
   /** Update the cache synchronously.
-    *
-    * In face of multiple in-flight updates competing for the `key` (see [[putAsync()]]),
-    * this method updates the cache only if the to-be-inserted tuple is the most recent
-    * (i.e. it has `validAt` highest amongst the competing updates).
     *
     * @param key the key at which to update the cache
     * @param validAt ordering discriminator for pending updates for the same key
     * @param value the value to insert
     */
-  def put(key: K, validAt: Long, value: V): Unit = Timed.value(
-    registerUpdateTimer, {
-      pendingUpdates.synchronized {
-        val competingLatestForKey =
-          pendingUpdates
-            .get(key)
-            .map { pendingUpdate =>
-              val oldLatestValidAt = pendingUpdate.latestValidAt
-              pendingUpdate.latestValidAt = Math.max(validAt, pendingUpdate.latestValidAt)
-              oldLatestValidAt
-            }
-            .getOrElse(Long.MinValue)
-
-        if (competingLatestForKey < validAt) cache.put(key, value) else ()
-      }
-    },
-  )
+  def put(key: K, validAt: Long, value: V)(implicit loggingContext: LoggingContext): Unit =
+    Timed.value(
+      registerUpdateTimer, {
+        pendingUpdates.synchronized {
+          // The mutable contract state cache update stream should generally increase the cacheIndex strictly monotonic.
+          // However, updates can be replayed in case of failure of the update stream.
+          // In this case, we must skip over the already seen updates (i.e. that have the validAt before or at the cacheIndex).
+          if (validAt > cacheIndex) {
+            pendingUpdates
+              .get(key)
+              .foreach(_.latestValidAt = validAt)
+            cacheIndex = validAt
+            putInternal(key, value, validAt)
```

Comment:
After internalizing the `cacheIndex`, we are safe to assume that the update here is the latest. No need to check against pending updates. Simplified